### PR TITLE
IGNITE-18345 SQL. Fix usage of java Date/Time API classes as parameters for query

### PR DESCRIPTION
**Pull request by zstan**

### CODE REVIEW

**Review by AMashenkov — APPROVED** [review]

**Comment by zstan** [comment]
@ygerzhedovich probably also you can check it ? thanks.

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDynamicParameterTest.java [hunk 1]
```diff
@@ -59,19 +64,28 @@ public void dropTables() {
...
-    void testMetadataTypesForDynamicParameters(ColumnType type) {
-        Object param = generateValueByType(RND.nextInt(), type);
-
-        assertQuery("SELECT typeof(?)").withParams(param).returns(toSqlType(type)).check();
-        assertQuery("SELECT ?").withParams(param).returns(param).columnMetadata(new MetadataMatcher().type(type)).check();
+    @Test
+    void testMetadataTypesForDynamicParameters() {
+        //    https://issues.apache.org/jira/browse/IGNITE-18789
+        //    https://issues.apache.org/jira/browse/IGNITE-18414
+        String[] excludedTypes = {"DECIMAL", "NUMBER", "BITMASK", "DURATION", "PERIOD"};
```
Comment (ygerzhedovich) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDynamicParameterTest.java:71]: Why is not the previous approach with @EnumSource ?

Comment (zstan) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDynamicParameterTest.java:71]: cause it start-stop ignite instances for each type and consume huge time.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/type/IgniteTypeFactory.java [hunk 2]
```diff
@@ -72,6 +74,26 @@ public class IgniteTypeFactory extends JavaTypeFactoryImpl {
...
+    /** Contains java types internally mapped into appropriate rel types. */
+    private static final Map<Class<?>, Supplier<RelDataType>> implementedJavaTypes = new IdentityHashMap<>();
+
+    {
+        {
+            implementedJavaTypes.put(LocalDate.class, () ->
+                    createTypeWithNullability(createSqlType(SqlTypeName.DATE), true));
+            implementedJavaTypes.put(LocalTime.class, () ->
+                    createTypeWithNullability(createSqlType(SqlTypeName.TIME), true));
+            implementedJavaTypes.put(LocalDateTime.class, () ->
```
[the remark this answers is not in the document]
Comment (zstan) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/type/IgniteTypeFactory.java:86]: I suppose we already have such a tests.

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDynamicParameterTest.java (hunk 1 above)
Comment (korlov42) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDynamicParameterTest.java:71]: you just removed the symptoms while the problem still exists

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/TypeUtils.java [hunk 3]
```diff
@@ -89,6 +91,33 @@ public class TypeUtils {
             SqlTypeName.INTERVAL_YEAR_MONTH
     );
 
+    private Set<Class<?>> correctParamClasses;
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/TypeUtils.java:94]: nitpicking: I would prefer to rename it to `supported`

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/TypeUtils.java [hunk 4]
```diff
@@ -89,6 +91,33 @@ public class TypeUtils {
...
+            correctParamClasses = Arrays.stream(ColumnType.values()).map(ColumnType::columnTypeToClass).collect(Collectors.toSet());
+
+            correctParamClasses.add(boolean.class);
+            correctParamClasses.add(byte.class);
+            correctParamClasses.add(short.class);
+            correctParamClasses.add(int.class);
+            correctParamClasses.add(long.class);
+            correctParamClasses.add(float.class);
+            correctParamClasses.add(double.class);
+            correctParamClasses.add(char[].class);
```
[the remark this answers is not in the document]
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/TypeUtils.java:107]: the same question for `char`

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/type/IgniteTypeFactory.java [hunk 5]
```diff
@@ -72,6 +74,26 @@ public class IgniteTypeFactory extends JavaTypeFactoryImpl {
     /** A registry that contains custom data types. **/
     private final CustomDataTypes customDataTypes;
 
+    /** Contains java types internally mapped into appropriate rel types. */
+    private static final Map<Class<?>, Supplier<RelDataType>> implementedJavaTypes = new IdentityHashMap<>();
```
[the remark this answers is not in the document]
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/type/IgniteTypeFactory.java:78]: why supplier as value?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/TypeUtils.java [hunk 6]
```diff
@@ -89,6 +91,33 @@ public class TypeUtils {
...
+            correctParamClasses.add(char.class);
+        }
+    }
+
+    public static TypeUtils instance() {
+        return INSTANCE;
```
[the remark this answers is not in the document]
Comment (zstan) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/TypeUtils.java:113]: whats the problem with initialization on demand ? If dyn params is not touched - no correctParamInstance instance is initialized.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/TypeUtils.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (zstan) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/TypeUtils.java]: ok, a change it into _java.lang.Character_  for supporting case like:
`assertQuery("SELECT COALESCE(null, ?)").withParams('j').returns("j").check();`

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDynamicParameterTest.java (hunk 1 above)
Comment (zstan) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDynamicParameterTest.java:71]: ok, i revert it

**Comment by zstan** [comment]
@AMashenkov @korlov42  thanks for suggestions, fixed.

**Comment by zstan** [comment]
@lowka seems your approach didn`t work for some corner cases, _ItDynamicParameterTest#testTrickyCase_ probably we need additional investigation here, can we move forward with current implementation?

**Comment by lowka** [comment]
> @lowka seems your approach didn`t work for some corner cases, _ItDynamicParameterTest#testTrickyCase_ probably we need additional investigation here, can we move forward with current implementation?

Sure.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/TypeUtils.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/TypeUtils.java]: Why do we forbid implicit casting from java.sql.Date to LocalDate, but allow implicit casting from char to string?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/TypeUtils.java [hunk 7]
```diff
@@ -89,6 +102,21 @@ public class TypeUtils {
...
 
+    private Set<Class<?>> supportedParamClasses;
+
+    private static class LazyHolder {
+        static final TypeUtils INSTANCE = new TypeUtils();
+    }
+
+    private static TypeUtils instance() {
+        return LazyHolder.INSTANCE;
+    }
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/TypeUtils.java:113]:
```suggestion
    private Set<Class<?>> supportedParamClasses;

    private static class SupportedParamClassesHolder {
        static final Set<Class<?>> supportedParamClasses;

        private static {
            supportedParamClasses = Arrays.stream(ColumnType.values()).map(ColumnType::columnTypeToClass).collect(Collectors.toSet());

            supportedParamClasses.add(boolean.class);
            supportedParamClasses.add(byte.class);
            supportedParamClasses.add(short.class);
            supportedParamClasses.add(int.class);
            supportedParamClasses.add(long.class);
            supportedParamClasses.add(float.class);
            supportedParamClasses.add(double.class);
            supportedParamClasses.add(Character.class);
        }
    }
```

**Pull request closed without merging**